crates/algebra/src/monoid.rs

```diff
@@ -0,0 +1,26 @@
+use crate::semigroup::Semigroup;
+
+/// A [`Monoid`] is an algebraic structure with [`Semigroup`] properties as well as equipped with
+/// a single identity element. The key property of the [`Monoid::empty`] value is as follows:
+///
+/// a: T
+/// T: Monoid
+/// T::empty().merge(a) == a.merge(T::empty()) == a
+///
+/// Intuitively this means that giving [`Monoid::empty`] as an argument to [`Semigroup::merge`] MUST
+/// return the merge's other argument.
+pub trait Monoid: Semigroup {
+    fn empty() -> Self;
+}
+
+/// The free catamorphism (fold) over a Monoidal iterator. If the iterator has no elements, it
+/// returns [`Monoid::empty`], otherwise it folds the iterator using [`Semigroup::merge`].
+pub fn concat<T: Monoid>(xs: impl IntoIterator<Item = T>) -> T {
+    xs.into_iter().fold(T::empty(), <T as Semigroup>::merge)
+}
+
+/// The universal catamorphism over all iterators. If given the iterant type has a morphism into a
```

Comment:
```suggestion
/// The universal catamorphism over all iterators. If the given iterant type has a morphism into a
```
?

Reply:
This function can be used to express all catamorphisms over any iterator, provided that you can project it to a Monoid.

Reply:
This comment was more about grammar. Is `If given the iterant type has a morphism...` correct?

Reply:
Yes. The "iterant type" is the `T` in `Item = T`.

Reply:
Okay I think what you meant. I'd still prefer something like: 
```suggestion
/// The universal catamorphism over all iterators. Provided that the iterant type has a morphism into a
```

Reply:
Oh I missed what you were suggesting before. You were correct in your original suggestion but I have changed it to your newest suggestion.